modules/ocf/files/ssl/ocf-lets-encrypt

```diff
@@ -0,0 +1,253 @@
+       web server. In this case, we start a temporary server (Python's
+       http.server) to serve the challenge files while making requests
+       and shut it down when we're finished.
+
+It is also possible for us to hit our rate limit for Let's Encrypt
+requests, but our limit as of writing is 1000 new certs per week with
+unlimited renewals, so this is only possible in case of catastrophic
+failure (i.e. we lose all of our web vhost certs).
+
+Once the needed certs are acquired, the corresponding .pem and .bundle
+files still have to be generated. This is left to the caller, i.e.
+puppet or a host-specific script.
+
+The return code indicates whether the script did anything, so the caller
+can reload services as necessary:
+
+    0       nothing changed
+    255     certs added/changed
+    else    error
+"""
+import argparse
+import functools
+import os
+import re
+import shutil
+import socket
+import subprocess
+import sys
+import tempfile
+from collections import namedtuple
+from contextlib import contextmanager
+from datetime import datetime
+from datetime import timedelta
+from datetime import timezone
+from pathlib import Path
+
+import dateutil.parser
+
+
+RENEW_WHEN_TIME_REMAINING = timedelta(days=45)
+
+DEFAULT_CERT_STORAGE_DIR = '/etc/ssl/private/'
+LE_ACCOUNT_KEY = Path('/etc/ssl/lets-encrypt/le-account.key')
+ACME_BASE_DIR = Path('/var/lib/lets-encrypt/')
+
+
+Request = namedtuple('Request', ('domain', 'cert_path', 'has_cert'))
+
+
+def debug(*args, **kwargs):
+    pass
+
+
+def expiration_date(cert_path):
+    output = subprocess.check_output((
+        'openssl', 'x509', '-enddate', '-noout',
+        '-in', str(cert_path),
+    ))
+    m = re.match(b'notAfter=(.+)$', output)
+    assert m, output
+    return dateutil.parser.parse(m.group(1).decode('ascii'))
+
+
+@contextmanager
+def make_csr(domain, key):
+    _, csr_path = tempfile.mkstemp()
+    try:
+        subprocess.check_call((
+            'openssl', 'req', '-new', '-sha256',
+            '-key', key,
+            '-subj', '/CN={}'.format(domain),
+            '-out', csr_path,
+        ))
+        yield csr_path
+    finally:
+        os.remove(csr_path)
+
+
+def acme_tiny(domain, key):
+    with make_csr(domain, key) as csr_path:
+        # TODO: use subprocess.run when we get python3.5
+        proc = subprocess.Popen(
+            (
+                'acme-tiny',
+                '--account-key', str(LE_ACCOUNT_KEY),
+                '--csr', csr_path,
+                '--acme-dir', str(ACME_BASE_DIR / '.well-known' / 'acme-challenge'),
+                # '--ca', 'https://acme-staging.api.letsencrypt.org',
+            ),
+            stdout=subprocess.PIPE,
+            stderr=subprocess.PIPE,
+        )
+        stdout, stderr = proc.communicate()
+    return stdout, stderr, proc.returncode
+
+
+def write_cert(cert, cert_path):
+    """Write certificate file for a domain."""
+    _, temp_cert_path = tempfile.mkstemp(dir=str(cert_path.parent), prefix='.tmp')
```

Comment:
Try using a `NamedTemporaryFile` with `delete=False` in the constructor instead of opening the temporary file twice.